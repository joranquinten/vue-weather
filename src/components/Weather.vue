<template>
  <div class="weather">
    Weather widget
    <div v-if="notAvailable">
      Sorry, localized weather forecast is currently not available
    </div>
    <Loading message="Loading..." v-if="loading"></Loading>
    <div v-if="weather && !loading">
      <h3>{{ weather.summary }}</h3>
      <p v-if="!weather.daily">{{ `${weather.temperature} &deg;C` }}</p>
      <WeatherDaily v-if="weather.daily" v-bind:daily="weather.daily" />
      <p class="last-updated">
        Last updated: {{ weather.time | formatTimeRelative }} geleden
      </p>

      <div class="type-switch">
        <router-link to="/">Nu</router-link>
        <router-link to="/daily">Komende week</router-link>
      </div>

      <LocationPicker
        v-bind:location="location"
        @location-change="handleUpdateLocation"
      />
    </div>
  </div>
</template>

<script>
import axios from 'axios'
import { distanceInWordsToNow } from 'date-fns'
import nl from 'date-fns/locale/nl'

import WeatherDaily from './WeatherDaily'
import LocationPicker from './LocationPicker'

let cache

const getWeatherForecast = async (location, type) => {
  const secret = process.env.VUE_APP_DARKSKY_API_KEY
  const endpoint = `https://api.darksky.net/forecast/${secret}/REPLACE_WITH_LAT,REPLACE_WITH_LON?lang=nl&units=ca&exclude=minutely,hourly,alerts,flags`

  const localForecast = endpoint
    .replace('REPLACE_WITH_LAT', location.latitude)
    .replace('REPLACE_WITH_LON', location.longitude)

  // TODO: Should check whether exists in collection of cache (array filter)
  if (
    cache &&
    (cache.latitude === location.latitude &&
      cache.longitude === location.longitude)
  ) {
    // TODO: Check whether to invalidate cache
    return weatherFormatter(cache, type)
  }

  return axios
    .get(localForecast)
    .then(response => {
      // TODO: Should push to cache for location
      cache = { ...response.data, location }
      return weatherFormatter(response.data, type)
    })
    .catch(error => console.error(error))
}

const weatherFormatter = (weather, type) => {
  if (!type) {
    return {
      summary: weather.currently.summary,
      icon: weather.currently.icon,
      latitude: weather.latitude,
      longitude: weather.longitude,
      temperature: weather.currently.temperature,
      time: weather.currently.time * 1000
    }
  }

  if (type === 'daily') {
    return {
      summary: weather.currently.summary,
      icon: weather.currently.icon,
      latitude: weather.latitude,
      longitude: weather.longitude,
      daily: weather.daily.data.reduce((report, day) => {
        report.push({
          summary: day.summary,
          time: day.time * 1000,
          temperatureHigh: day.temperatureHigh,
          temperatureLow: day.temperatureLow,
          icon: day.icon
        })
        return report
      }, []),
      time: weather.currently.time * 1000
    }
  }
}

export default {
  name: 'Weather',
  data() {
    return {
      weather: null,
      location: null,
      loading: null,
      notAvailable: null
    }
  },
  components: {
    WeatherDaily,
    LocationPicker
  },
  filters: {
    formatTimeRelative(value) {
      if (!value) return ''
      return distanceInWordsToNow(new Date(value), { locale: nl })
    }
  },
  watch: {
    $route(to, from) {
      getWeatherForecast(this.location, this.$route.params.forecastType).then(
        weather => {
          this.weather = weather
          this.loading = false
          return this
        }
      )
    }
  },
  mounted() {
    if ('geolocation' in navigator) {
      this.loading = true
      navigator.geolocation.getCurrentPosition(position => {
        this.location = {
          latitude: position.coords.latitude,
          longitude: position.coords.longitude
        }

        getWeatherForecast(this.location, this.$route.params.forecastType).then(
          weather => {
            this.weather = weather
            this.loading = false
          }
        )
      })
    } else {
      return (this.notAvailable = true)
    }
  },
  methods: {
    handleUpdateLocation(value) {
      this.loading = true
      getWeatherForecast(value, this.$route.params.forecastType).then(
        weather => {
          this.weather = weather
          this.loading = false
        }
      )
    }
  }
}
</script>

<style scoped>
.weather {
  border: 1px solid #333;
}
</style>
