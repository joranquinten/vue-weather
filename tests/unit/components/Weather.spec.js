import { shallowMount, mount } from '@vue/test-utils'
// import mockNow from 'jest-mock-now'
import flushPromises from 'flush-promises'
import Weather from '../../../src/components/Weather'

// TODO: Stub components (Loading, WeatherDaily)

const mockGeolocation = {
  getCurrentPosition: success =>
    success({
      coords: { latitude: 1, longitude: 2 }
    })
}

const mockComponents = {
  stubs: {
    Loading: "<div class='stub'>Loading</div>",
    WeatherDaily: "<div class='stub'>WeatherDaily</div>",
    'router-link': "<span class='stub'>RouterLink</span>"
  }
}

jest.mock('axios', () => {
  return {
    get: () =>
      Promise.resolve({
        data: {
          latitude: 51.613358399999996,
          longitude: 5.5310954,
          currently: {
            time: 1555403544,
            summary: 'Licht bewolkt',
            icon: 'partly-cloudy-day',
            temperature: 12.91
          },
          daily: {
            summary:
              'De hele week geen neerslag met temperaturen stijgend tot 26Â°C volgende dinsdag.',
            icon: 'clear-day',
            data: [
              {
                summary: 'Dag 1',
                time: 1555365600000,
                temperatureHigh: 20.41,
                temperatureLow: 8.57,
                icon: 'icon-day-one'
              },
              {
                summary: 'Dag 2',
                time: 1555452000000,
                temperatureHigh: 10.17,
                temperatureLow: -1.24,
                icon: 'icon-day-two'
              }
            ]
          }
        }
      })
  }
})

// console.log(mockNow(new Date('2019-01-31')))
describe('Weather.vue', () => {
  it('renders without crashing', () => {
    const wrapper = shallowMount(Weather, mockComponents)
    expect(wrapper).toMatchSnapshot()
  })

  it('should display availability status based on presence of geolocation', () => {
    // Remove geolocation
    delete global.navigator.geolocation

    const wrapperWithoutGeoLocation = mount(Weather, {
      ...mockComponents,
      mocks: {
        $route: {
          params: { forecastType: '' }
        }
      }
    })
    expect(wrapperWithoutGeoLocation.vm.notAvailable).toEqual(true)

    // Restore geolocation
    global.navigator.geolocation = mockGeolocation

    const wrapperWithGeoLocation = mount(Weather, {
      ...mockComponents,
      mocks: {
        $route: {
          params: { forecastType: '' }
        }
      }
    })
    expect(wrapperWithGeoLocation.vm.notAvailable).toEqual(null)
  })

  it('should update the loading state when a request is made', async () => {
    const wrapper = mount(Weather, {
      ...mockComponents,
      mocks: {
        $route: {
          params: { forecastType: '' }
        }
      }
    })

    wrapper.vm.handleUpdateLocation({ latitude: 1, longitude: 2 })
    expect(wrapper.vm.loading).toEqual(true)
    await flushPromises()

    expect(wrapper.vm.loading).toEqual(false)
    expect(wrapper.vm.location).toEqual({ latitude: 1, longitude: 2 })

    expect(wrapper.vm.weather).toEqual({
      summary: 'Licht bewolkt',
      icon: 'partly-cloudy-day',
      latitude: 51.613358399999996,
      longitude: 5.5310954,
      temperature: 12.91,
      time: 1555403544000
    })
  })
})
