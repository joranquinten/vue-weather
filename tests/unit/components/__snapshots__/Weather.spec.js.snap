// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather.vue renders without crashing 1`] = `
<div class="weather">
  Weather widget
  <!---->
  <!---->
  <!---->
</div>
`;
